### ISSUE TITLE
3D bounding box generation seems suboptimal

### ISSUE THREAD

**Issue by FabianIsensee** [JunMa11/CVPR-MedSegFMCompetition#5]
Creating the bounding box based on the xy extend of the segmentation at the z_mid slice causes weird and honestly pretty unrealistic bboxes to be generated, for example:

![Image](https://github.com/user-attachments/assets/63de2b80-47e9-4d3c-8983-2011e6b228ee)

This is liver079 from LiTS, first image I tested.

Code for reproducibility:
```python
from typing import Optional, Tuple, List

import SimpleITK
import numpy as np
import torch

from nnunetv2.imageio.simpleitk_reader_writer import SimpleITKIO
from nnunetv2.training.interaction_simulation.interactions.bounding_boxv2 import BoundingBoxInteractionv2


def mask2D_to_bbox(gt2D):
    y_indices, x_indices = np.where(gt2D > 0)
    x_min, x_max = np.min(x_indices), np.max(x_indices)
    y_min, y_max = np.min(y_indices), np.max(y_indices)
    # add perturbation to bounding box coordinates
    H, W = gt2D.shape
    bbox_shift = np.random.randint(0, 6, 1)[0]
    scale_y, scale_x = gt2D.shape
    bbox_shift_x = int(bbox_shift * scale_x/256)
    bbox_shift_y = int(bbox_shift * scale_y/256)
    #print(f'{bbox_shift_x=} {bbox_shift_y=} with orig {bbox_shift=}')
    x_min = max(0, x_min - bbox_shift_x)
    x_max = min(W-1, x_max + bbox_shift_x)
    y_min = max(0, y_min - bbox_shift_y)
    y_max = min(H-1, y_max + bbox_shift_y)
    boxes = np.array([x_min, y_min, x_max, y_max])
    return boxes


def mask3D_to_bbox(gt3D):
    b_dict = {}
    z_indices, y_indices, x_indices = np.where(gt3D > 0)
    z_min, z_max = np.min(z_indices), np.max(z_indices)
    # middle of z_indices
    z_middle = z_indices[len(z_indices)//2]
    D, H, W = gt3D.shape
    b_dict['z_min'] = z_min
    b_dict['z_max'] = z_max
    b_dict['z_mid'] = z_middle

    gt_mid = gt3D[z_middle]

    box_2d = mask2D_to_bbox(gt_mid)
    x_min, y_min, x_max, y_max = box_2d
    b_dict['z_mid_x_min'] = x_min
    b_dict['z_mid_y_min'] = y_min
    b_dict['z_mid_x_max'] = x_max
    b_dict['z_mid_y_max'] = y_max

    assert z_min == max(0, z_min)
    assert z_max == min(D-1, z_max)
    return b_dict


file = '/home/isensee/drives/E132-Rohdaten/nnUNetv2/Dataset003_Liver/labelsTr/liver_79.nii.gz'
seg_itk = SimpleITK.ReadImage(file)
seg = (SimpleITK.GetArrayFromImage(seg_itk) > 0).astype(np.uint8)
bbox = mask3D_to_bbox(seg)
bbox_img = np.zeros_like(seg)
bbox_img[
    bbox['z_min']: bbox['z_max'] + 1,
    bbox['z_mid_y_min']: bbox['z_mid_y_max'] + 1,
    bbox['z_mid_x_min']: bbox['z_mid_x_max'] + 1
] = 1
bbox_itk = SimpleITK.GetImageFromArray(bbox_img)
bbox_itk.CopyInformation(seg_itk)
SimpleITK.WriteImage(bbox_itk, '/home/isensee/temp/liver_79_bbox.nii.gz')
```

Not sure if you would consider changing that at this point but this seems like an odd design choice.
Best,
Fabian

**Comment by FabianIsensee** [comment]
Ah and it would maybe be good to clearly communicate that the bounding boxes here are not half open interval [min, max) but closed interval [min, max] :-)

**Comment by JunMa11** [comment]
Hi @FabianIsensee ,

Thanks for your feedback very much. 

This is intentional because it's not user-friendly to directly draw a 3D bounding box in typical image viewers.  

As an alternative, our box prompt strategy is to draw a 2D bounding box in the middle slice followed by specifying the 1st slice and last slice of the target. We have implemented this strategy in 3D slicer and found it significantly more practical compared to the traditional 3D bounding box approach. 

Since we'll work with the top teams to incorporate their algorithms into 3D slicer to facilitate real-world applications, this design will not be changed for better utility. 

We'll also clarify the closed interval in the dataset readme file. 

Remark: 
1. the initial bounding box prompt is optional to use. Participants can also directly start with point prompts and our evaluation code will simulate an empty mask for the box prompt in this case.  
2. some objects don’t have box prompt, e.g., vessels

**Comment by FabianIsensee** [comment]
Hey @JunMa11, I am aware of this strategy to draw a 3D bbox and would challenge that this is a good balance between precision and usability. If you really want 3D bounding boxes, a more reliable way to generate them would be to ask the user to place 6 points, one at each border (zlow, zhigh, ylow, yhigh etc) of the structure. 
Not using the bbox would give a competitive disadvantage so I think everyone is incentivized to use it, even if it may be unreliable.
Best,
Fabian

**Comment by JunMa11** [comment]
Hi Fabian,

Thanks for your great suggestion on 3D box. We also tried this way with our annotation team by asking them place extreme points in 3D slicer. However, none of them likes this way because it is time-consuming to find the extreme points for xmin, xmax, ymin, and ymax. 
After comparing several strategies, we found the current strategy is the best trade-off between user-friendliness and precision.
